arduino-ide-extension/src/browser/theia/workspace/workspace-commands.ts

```diff
@@ -67,51 +91,72 @@ export class WorkspaceCommandContribution extends TheiaWorkspaceCommandContribut
-    if (nameWithExt) {
-      const fileUri = parentUri.resolve(nameWithExt);
-      await this.fileService.createFile(fileUri);
-      this.fireCreateNewFile({ parent: parentUri, uri: fileUri });
-      open(this.openerService, fileUri);
+    const name = await this.openDialog(dialog, parentUri);
+    if (!name) {
+      return;
     }
+    const nameWithExt = this.maybeAppendInoExt(name);
+    const fileUri = parentUri.resolve(nameWithExt);
+    await this.fileService.createFile(fileUri);
+    this.fireCreateNewFile({ parent: parentUri, uri: fileUri });
+    open(this.openerService, fileUri);
   }
 
   protected override async validateFileName(
-    name: string,
+    userInput: string,
     parent: FileStat,
     recursive = false
   ): Promise<string> {
-    // In the Java IDE the followings are the rules:
-    //  - `name` without an extension should default to `name.ino`.
-    //  - `name` with a single trailing `.` also defaults to `name.ino`.
-    const nameWithExt = this.maybeAppendInoExt(name);
-    const errorMessage = await super.validateFileName(
-      nameWithExt,
-      parent,
-      recursive
-    );
+    // If name does not have extension or ends with trailing dot (from IDE 1.x), treat it as an .ino file.
+    // If has extension,
+    //  - if unsupported extension -> error
+    //  - if has a code file extension -> apply folder name validation without the extension and use the Theia-based validation
+    //  - if has any additional file extension -> use the default Theia-based validation
+    const fileInput = parseFileInput(userInput);
+    const { name, extension } = fileInput;
+    if (!Sketch.Extensions.ALL.includes(extension)) {
+      return invalidExtension(extension);
+    }
+    let errorMessage: string | undefined = undefined;
+    if (Sketch.Extensions.CODE_FILES.includes(extension)) {
+      errorMessage = this._validationContext?.isCloud
+        ? Sketch.validateCloudSketchFolderName(name)
+        : Sketch.validateSketchFolderName(name);
+    }
     if (errorMessage) {
-      return errorMessage;
-    }
-    const extension = nameWithExt.split('.').pop();
-    if (!extension) {
-      return nls.localize(
-        'theia/workspace/invalidFilename',
-        'Invalid filename.'
-      ); // XXX: this should not happen as we forcefully append `.ino` if it's not there.
-    }
-    if (Sketch.Extensions.ALL.indexOf(`.${extension}`) === -1) {
-      return nls.localize(
-        'theia/workspace/invalidExtension',
-        '.{0} is not a valid extension',
-        extension
-      );
+      return this.maybeRemapAlreadyExistsMessage(errorMessage, userInput);
+    }
+    errorMessage = await super.validateFileName(userInput, parent, recursive); // run the default Theia validation with the raw input.
+    if (errorMessage) {
+      return this.maybeRemapAlreadyExistsMessage(errorMessage, userInput);
+    }
+    // It's a legacy behavior from IDE 1.x. Validate the file it were an `.ino` file.
```

Comment:
```suggestion
    // It's a legacy behavior from IDE 1.x. Validate the file as if it were an `.ino` file.
```

Correction of the wording. I believe this was the intended meaning of the sentence.

Reply:
b10d42b